File: CloneGraph.java
class Solution {
    Node cloneGraph(Node node) {
        // code here
        Queue<Node> q = new LinkedList<>();
        q.add(node);
        Map<Node, Node> map = new HashMap<>();
        map.put(node, new Node(node.val));
        while(!q.isEmpty()){
            Node u = q.poll();
            Node clonedU = map.get(u);
            for(Node v : u.neighbors){
                Node clonedV = map.get(v);
                if(clonedV == null){
                    clonedV = new Node(v.val);
                    q.add(v);
                    map.put(v, clonedV);
                }
                clonedU.neighbors.add(clonedV);
            }
        }
        return map.get(node);
    }
}

File: CloneGraph.cpp
class Solution {
  public:
    Node* cloneGraph(Node* node) {
        // code here
        if (!node) return nullptr;

        unordered_map<Node*, Node*> map;
        queue<Node*> q;
        q.push(node);
        map[node] = new Node(node->val);
    
        while (!q.empty()) {
            Node* u = q.front();
            q.pop();
    
            for (Node* v : u->neighbors) {
                if (map.find(v) == map.end()) {
                    map[v] = new Node(v->val);
                    q.push(v);
                }
                map[u]->neighbors.push_back(map[v]);
            }
        }
    
        return map[node];
    }
};
